File: server/internal/process/gamescope.go
package process

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/ametis70/launchscope/server/internal/apps"
)

// UIArgv builds the argv for launching the launchscope UI.
// launchscopeBin may be a single binary name ("launchscope") or a
// space-separated command ("love /path/to/ui") — it is split like an exec string.
// Note: LAUNCHSCOPE_MODE=server is injected as an environment variable by
// launchUI(), not here.
func UIArgv(launchscopeBin string) ([]string, error) {
	return splitExec(launchscopeBin)
}

// AppArgv builds the argv slice for launching a configured app.
// When gamescope is disabled the exec string is split directly.
// The schema mirrors lib/gamescope.lua in the UI.
func AppArgv(app *apps.App) ([]string, error) {
	gs := app.Gamescope

	if !gs.Enabled {
		return splitExec(app.Exec)
	}

	// Resolve output and inner resolutions.
	ow := gs.Output.Width
	oh := gs.Output.Height
	orr := gs.Output.Refresh
	if ow == 0 {
		ow = 1920
	}
	if oh == 0 {
		oh = 1080
	}
	if orr == 0 {
		orr = 60
	}

	iw := gs.Inner.Width
	ih := gs.Inner.Height
	if iw == 0 {
		iw = ow
	}
	if ih == 0 {
		ih = oh
	}

	args := []string{"gamescope"}

	if gs.Fullscreen {
		args = append(args, "-f")
	}

	args = append(args,
		"-W", itoa(ow),
		"-H", itoa(oh),
		"-w", itoa(iw),
		"-h", itoa(ih),
		"-r", itoa(orr),
	)

	// Upscaler filter and scaler.
	if gs.Filter != "" {
		scaler := gs.Scaler
		// Only auto-select a scaler when resolutions actually differ.
		if scaler == "" && (iw != ow || ih != oh) {
			scaler = "fit"
		}
		// Only pass -S when a scaler is meaningful (filter set and either an
		// explicit scaler was requested or resolutions differ).
		if scaler != "" {
			args = append(args, "-S", scaler)
		}
		args = append(args, "-F", gs.Filter)
	}

	if gs.Sharpness != nil {
		args = append(args, "--sharpness", itoa(*gs.Sharpness))
	}

	if gs.HDR {
		args = append(args, "--hdr-enabled")
	}
	if gs.AdaptiveSync {
		args = append(args, "--adaptive-sync")
	}
	if gs.ForceGrab {
		args = append(args, "--force-grab-cursor")
	}
	if gs.ExposeWayland {
		args = append(args, "--expose-wayland")
	}
	if gs.CompositeDebug {
		args = append(args, "--composite-debug")
	}
	if gs.MangoApp {
		args = append(args, "--mangoapp")
	}

	// Validate and append extra flags. Reject any element equal to "--" to
	// prevent injection that would move the app exec to the gamescope flags
	// side of the separator.
	for _, f := range gs.ExtraFlags {
		if f == "--" {
			return nil, fmt.Errorf("ExtraFlags must not contain \"--\" (app id %q)", app.ID)
		}
		args = append(args, f)
	}

	args = append(args, "--")

	execArgs, err := splitExec(app.Exec)
	if err != nil {
		return nil, fmt.Errorf("parsing exec for app %q: %w", app.ID, err)
	}
	args = append(args, execArgs...)

	return args, nil
}

// splitExec splits a shell-style command string into argv.
// Handles single/double quoted tokens; no variable expansion.
// Returns an error if the input contains an unclosed quote.
func splitExec(s string) ([]string, error) {
	var args []string
	var cur strings.Builder
	inQ := false
	qChar := rune(0)

	for _, r := range s {
		switch {
		case inQ && r == qChar:
			inQ = false
		case !inQ && (r == '\'' || r == '"'):
			inQ = true
			qChar = r
		case !inQ && r == ' ':
			if cur.Len() > 0 {
				args = append(args, cur.String())
				cur.Reset()
			}
		default:
			cur.WriteRune(r)
		}
	}

	if inQ {
		return nil, fmt.Errorf("unclosed quote in command string: %q", s)
	}

	if cur.Len() > 0 {
		args = append(args, cur.String())
	}
	return args, nil
}

func itoa(n int) string { return strconv.Itoa(n) }

File: ui/lib/gamescope.lua
-- Gamescope session configuration.
-- Shared by both the launcher display config and per-app launch config.
--
-- Schema (all fields optional):
--   enabled        bool     wrap in gamescope (default true for apps; N/A for launcher)
--   fullscreen     bool     -f flag (default true)
--   output         table    { width, height, refresh }  →  -W -H -r
--   inner          table    { width, height }            →  -w -h  (defaults to output)
--   filter         string   upscaler filter: "linear" | "nearest" | "fsr" | "nis" | "pixel"
--                           →  -F <filter>  (also sets -S fit when inner != output)
--   scaler         string   scaler type: "auto" | "integer" | "fit" | "fill" | "stretch"
--                           →  -S <scaler>  (default "fit" when filter is set)
--   sharpness      int      0 (max) – 20 (min)  →  --sharpness N
--   hdr            bool     →  --hdr-enabled
--   adaptive_sync  bool     →  --adaptive-sync
--   force_grab     bool     →  --force-grab-cursor
--   expose_wayland bool     →  --expose-wayland
--   composite_debug bool    →  --composite-debug
--   mangoapp       bool     →  --mangoapp
--   extra_flags    [string] verbatim flags before --

local M = {}

local function shellQuote(s)
    return "'" .. tostring(s):gsub("'", "'\\''") .. "'"
end

-- Build a gamescope argv string from a gamescope config table.
-- If exec is provided, appends " -- <exec>" at the end.
function M.buildArgv(gs, exec)
    gs = gs or {}

    local out = gs.output or {}
    local inn = gs.inner or {}
    local ow = out.width or 1920
    local oh = out.height or 1080
    local orr = out.refresh or 60
    local iw = inn.width or ow
    local ih = inn.height or oh

    local parts = { "gamescope" }

    if gs.fullscreen ~= false then
        parts[#parts + 1] = "-f"
    end

    parts[#parts + 1] = "-W"
    parts[#parts + 1] = tostring(ow)
    parts[#parts + 1] = "-H"
    parts[#parts + 1] = tostring(oh)
    parts[#parts + 1] = "-w"
    parts[#parts + 1] = tostring(iw)
    parts[#parts + 1] = "-h"
    parts[#parts + 1] = tostring(ih)
    parts[#parts + 1] = "-r"
    parts[#parts + 1] = tostring(orr)

    -- Upscaler filter (-F) and scaler (-S).
    -- Only emit when inner != output (otherwise no upscaling is needed).
    local has_inner = (iw ~= ow or ih ~= oh)
    local filter = gs.filter
    if filter then
        local scaler = gs.scaler or (has_inner and "fit" or "auto")
        parts[#parts + 1] = "-S"
        parts[#parts + 1] = scaler
        parts[#parts + 1] = "-F"
        parts[#parts + 1] = filter
    end

    if gs.sharpness then
        parts[#parts + 1] = "--sharpness"
        parts[#parts + 1] = tostring(gs.sharpness)
    end

    if gs.hdr then
        parts[#parts + 1] = "--hdr-enabled"
    end
    if gs.adaptive_sync then
        parts[#parts + 1] = "--adaptive-sync"
    end
    if gs.force_grab then
        parts[#parts + 1] = "--force-grab-cursor"
    end
    if gs.expose_wayland then
        parts[#parts + 1] = "--expose-wayland"
    end
    if gs.composite_debug then
        parts[#parts + 1] = "--composite-debug"
    end
    if gs.mangoapp then
        parts[#parts + 1] = "--mangoapp"
    end

    for _, flag in ipairs(gs.extra_flags or {}) do
        parts[#parts + 1] = shellQuote(flag)
    end

    if exec then
        parts[#parts + 1] = "--"
        parts[#parts + 1] = exec
    end

    return table.concat(parts, " ")
end

M.DEFAULTS = {
    fullscreen = true,
    output = { width = 1920, height = 1080, refresh = 60 },
    inner = nil,
    filter = nil,
    scaler = nil,
    sharpness = nil,
    hdr = false,
    adaptive_sync = false,
    force_grab = false,
    expose_wayland = false,
    composite_debug = false,
    mangoapp = false,
    extra_flags = {},
}

return M
